### ISSUE TITLE
Replace ape::mst by igraph::mst

### ISSUE THREAD

**Issue by joheli** [joheli/yea13#2]
In an effort to replace `ape::mst` by `igraph::mst` I notice that the function behave differently for 'real' data, i.e. data with zeroes.

## ape and igraph produce same minimum spanning topologies on random data
```
library(dplyr)
library(igraph)
library(ape)
library(tidyr)
source("dist_df.r")

# generate random matrix
set.seed(12)
mx <- matrix(runif(64), ncol = 8)

# generate dist object
d <- dist(mx)

# # 'real data'
# d <- s_aureus[base::sample(1:nrow(s_aureus), 20), ] %>% # select random subset
#   select(-id, -time, -unit) %>%
#   dist(method = "manhattan")
  
# Igraph
# transform dist object to data.frame and feed to igraph for graph and minimum spanning tree generation
mst.i <- dist_df(d) %>%
  igraph::graph_from_data_frame(directed = FALSE) %>%
  igraph::mst()

# Ape
# generate graph from ape::mst object
mst.a <- ape::mst(d) %>%
  igraph::graph_from_adjacency_matrix(mode = "undirected")

# Show side by side
split.screen(c(1,2))
screen(1)
plot(mst.i, main = "Igraph MST")
screen(2)
plot(mst.a, main = "Ape MST")
close.screen(all = TRUE)
```
![igraph_vs_ape_1](https://user-images.githubusercontent.com/152780/74020946-e7bdfa80-499a-11ea-8ce0-2035cebf123c.png)

## Topologies differ, however, with 'real' data

After uncommenting the block

```
# 'real data'
d <- s_aureus[base::sample(1:nrow(s_aureus), 20), ] %>% # select random subset
   select(-id, -time, -unit) %>%
   dist(method = "manhattan")
```
topologies differ substantially:

![igraph_vs_ape_2](https://user-images.githubusercontent.com/152780/74021050-1dfb7a00-499b-11ea-8686-67992e15b2ae.png)

What is going on? Is it because of zero distances in latter distance matrix?

**Comment by joheli** [comment]
p.s. here's the code of `dist_df`
```
dist_df <- function(d) {
  m <- as.matrix(d)
  dm <- as.data.frame(m)
  dm$from <- rownames(dm)
  tidyr::gather(dm, key = to, value = weight, -from)
}
```

**Comment by joheli** [comment]
### Packages `ape` and `igraph` handle Manhattan distances slightly differently

Inserting the following code 
```
set.seed(12)
n <- 12 # number of points
mx <- matrix(round(runif(n^2, 1, 3)), ncol = n)
d <- dist(mx, method = "manhattan")
```
into the script above yields slightly different topologies:
![igraph_vs_ape_3](https://user-images.githubusercontent.com/152780/74250348-c20e5980-4cea-11ea-8753-b6e48568facb.png)
Here, one could argue, which tree is actually better (beats me!).
### Packages `ape` and `igraph` are worlds apart when dealing with recycled values (i.e. zero distances)
Nevertheless, introducing zero distances by recycling a few nodes, as in the following code
```
set.seed(12)
n <- 12 # number of points
mx <- matrix(round(runif(n^2, 1, 3)), ncol = n)
ix <- sample(1:n, size = n, replace = TRUE)
mx <- mx[ix, ix] # recycle points, thereby generating zero distances
d <- dist(mx, method = "manhattan")
```
result in completely different topologies (as in first post). In my opinion, `ape` handles the situation much more constructively (for the purpose of generating 'Ypma dissimilarities') than `igraph`:
![igraph_vs_ape_4](https://user-images.githubusercontent.com/152780/74250648-347f3980-4ceb-11ea-8e2b-2920bd4f2d50.png)
In summary, Manhattan distances and recycled values lead to different minimum spanning trees in `ape` and `igraph`, respectively.

**Comment by joheli** [comment]
Just to underline above statement regarding "for the purpose of generating 'Ypma dissimilarities'" take a look a the different output by `igraph::distance`:
```
> distances(mst.i)
    1  2  3  4  5  6  7  8  9 10 11 12
1   0 10  7 11  7  7  0  7  4  4  0  8
2  10  0  3 21  3  3 10 17 14 14 10 18
3   7  3  0 18  0  0  7 14 11 11  7 15
4  11 21 18  0 18 18 11 18  7  7 11  3
5   7  3  0 18  0  0  7 14 11 11  7 15
6   7  3  0 18  0  0  7 14 11 11  7 15
7   0 10  7 11  7  7  0  7  4  4  0  8
8   7 17 14 18 14 14  7  0 11 11  7 15
9   4 14 11  7 11 11  4 11  0  0  4  4
10  4 14 11  7 11 11  4 11  0  0  4  4
11  0 10  7 11  7  7  0  7  4  4  0  8
12  8 18 15  3 15 15  8 15  4  4  8  0
> distances(mst.a)
   1 2 3 4 5 6 7 8 9 10 11 12
1  0 2 1 2 2 2 1 1 1  2  1  1
2  2 0 1 4 2 2 3 3 3  4  3  3
3  1 1 0 3 1 1 2 2 2  3  2  2
4  2 4 3 0 4 4 3 3 3  4  3  1
5  2 2 1 4 0 2 3 3 3  4  3  3
6  2 2 1 4 2 0 3 3 3  4  3  3
7  1 3 2 3 3 3 0 2 2  3  2  2
8  1 3 2 3 3 3 2 0 2  3  2  2
9  1 3 2 3 3 3 2 2 0  1  2  2
10 2 4 3 4 4 4 3 3 1  0  3  3
11 1 3 2 3 3 3 2 2 2  3  0  2
12 1 3 2 1 3 3 2 2 2  3  2  0
```
Again, result by `ape` is more favourable.

**Comment by pmenzel** [comment]
I also had some issues with 0 dists in my MST experiments with igraph, so I just replaced 0 dists with 0.1 dists in the inital distance matrix before calculating the MST 
```{r}

d <- dist.gene( ...)
m.dists <- as.matrix(d)

# HACK to avoid non-inclusion of 0-dist edges in the initial graph g
for (i in 1:dim(m.dists)[1]) {
  for (j in 1:dim(m.dists)[2]) {
    if (i == j) {                                                                                                                                                                                                                                
      next
    }
    if (m.dists[i, j] == 0) {
      m.dists[i, j] <- 0.1
    }
  }
}

g.dist <- igraph::graph_from_adjacency_matrix(m.dists, mode = "lower", diag = FALSE, weighted = TRUE)

E(g.dist)$label <- as.character(E(g.dist)$weight)

t.mst <- igraph::mst(g.dist)

# change values again for prettier printing 
t.mst <- set_edge_attr(t.mst, "weight", value = case_when(
    E(t.mst)$weight == 0.1 ~ 0.0001,
    E(t.mst)$weight == 1 ~ 0.85,
    TRUE ~ log(E(t.mst)$weight)
  ))                  

 plot <- ggraph(t.mst, layout = "igraph", algorithm = "kk") +
    geom_edge_fan(aes(label = label), color = "grey70", label_colour = "grey30") +
    geom_node_point(size = 10, color = "grey70") +
    geom_node_text(aes(label = name), repel = TRUE) +
    scale_color_discrete(guide = F) +
    theme_graph()

                     
```

**Comment by joheli** [comment]
Thank you, Peter!

You may be onto something there - I also created a 'hack' to add random values to zero distances (see function `zero.noise`). In fact it changes (imho improves) the result by `igraph::mst`. Interestingly, application of `zero.noise` does not change output of `ape::mst`, though.

Here's the experiment:

```
set.seed(12)
n <- 12 # number of points
mx <- matrix(round(runif(n^2, 1, 3)), ncol = n)
ix <- sample(1:n, size = n, replace = TRUE)
mx <- mx[ix, ix] # generate zero distances by recycling a few cases

d <- dist(mx, method = "manhattan")
d <- as.dist(zero.noise(as.matrix(d), lo = .5, hi = 1.5)) # <-- here's the hack
```
The rest is as above.
Above lines generate the following output:
![igraph_vs_ape_5](https://user-images.githubusercontent.com/152780/74743725-32742800-5261-11ea-9aa3-2d2a7aacc569.png)
The question now is: which one is 'right'? :smile:

**Comment by pmenzel** [comment]
The differences might be due to multiple possible optimal solutions, when the distances are sampled from a uniform distribution, so that each package chooses one of them.
I just tried with a normal distribution and get the same MSTs from both packages (sample size = 2 :) )

```
library(patchwork)
library(igraph)
library(ape)
library(tidyverse)
library(ggraph)

set.seed(112)

n <- 20

m <- matrix(nrow=n, ncol=n)
# use normally distributed distances
m[lower.tri(m)] <- round(rnorm(n*(n-1)/2, mean = 40, sd=30))
m[m<0] <- 0

#add noise
m[m==0 & !is.na(m)] <- runif(length(m[m==0 & !is.na(m)]), 0.5, 1.0)

rownames(m) <- 1:n
colnames(m) <- 1:n

mst.igraph <- igraph::graph_from_adjacency_matrix(m, mode = "lower", diag = FALSE, weighted = TRUE) %>%
  igraph::mst()

E(mst.igraph)$weight <- 1

p.igraph <- ggraph(mst.igraph, layout = "igraph", algorithm = "kk") +
  geom_edge_fan(color = "grey70", label_colour = "grey30") +
  geom_node_point(size = 7, color = "grey70", fill = "grey90",shape=21) +
  scale_color_discrete(guide = F) +
  geom_node_text(aes(label = name)) +
  ggtitle("igraph") +
  theme_graph()


mst.ape <- ape::mst(as.dist(m)) %>%
  igraph::graph_from_adjacency_matrix(mode = "undirected")

p.ape <- ggraph(mst.ape, layout = "igraph", algorithm = "kk") +
  geom_edge_fan(color = "grey70", label_colour = "grey30") +
  geom_node_point(size = 7, color = "grey70", fill = "grey90",shape=21) +
  scale_color_discrete(guide = F) +
  geom_node_text(aes(label = name)) +
  ggtitle("ape") +
  theme_graph()

p.igraph + p.ape
```
![Rplot](https://user-images.githubusercontent.com/1469600/74828479-79205b80-530f-11ea-8314-edb0cf7fd45c.png)

**Comment by joheli** [comment]
This is very enlightening - thank you so much!

I guess you´re placing your finger in one of the wounds: distances in my example are __not__ normally distributed; may this explain different topologies?

Looky here:

```
library(patchwork)
library(dplyr)
library(igraph)
library(ape)
library(tidyr)

# generate random matrix
set.seed(12)
n <- 12 # number of points
mx <- matrix(round(runif(n^2, 1, 3)), ncol = n)
ix <- sample(1:n, size = n, replace = TRUE)
mx <- mx[ix, ix] # generate zero distances by recycling a few cases

# generate dist object
d <- dist(mx, method = "manhattan")
dm <- as.matrix(d)
dm[dm == 0] <- runif(length(dm[dm==0]), 0.5, 1.0)

# Igraph
mst.i <- igraph::graph_from_adjacency_matrix(dm, mode = "lower", diag = FALSE, weighted = TRUE) %>%
  igraph::mst()

E(mst.i)$weight <- 1

# Ape
# generate graph from ape::mst object
mst.a <- ape::mst(d) %>%
  igraph::graph_from_adjacency_matrix(mode = "undirected")

# Plot
p.igraph <- ggraph(mst.i, layout = "igraph", algorithm = "kk") +
  geom_edge_fan(color = "grey70", label_colour = "grey30") +
  geom_node_point(size = 7, color = "grey70", fill = "grey90", shape=21) +
  scale_color_discrete(guide = F) +
  geom_node_text(aes(label = name)) +
  ggtitle("igraph") +
  theme_graph()

p.ape <- ggraph(mst.a, layout = "igraph", algorithm = "kk") +
  geom_edge_fan(color = "grey70", label_colour = "grey30") +
  geom_node_point(size = 7, color = "grey70", fill = "grey90", shape=21) +
  scale_color_discrete(guide = F) +
  geom_node_text(aes(label = name)) +
  ggtitle("ape") +
  theme_graph()

print(p.igraph + p.ape)
```
![igraph_vs_ape_6](https://user-images.githubusercontent.com/152780/75267414-800b0a80-57f6-11ea-992f-91c3ed9d0eef.png)

Distances are not normally distributed (as expected for manhattan distance):

```
> shapiro.test(dm[lower.tri(dm)])

	Shapiro-Wilk normality test

data:  dm[lower.tri(dm)]
W = 0.95033, p-value = 0.01021
```
We may yet solve the mystery!

**Comment by joheli** [comment]
Function `yea13::mstC` is a C++ transcript of `ape::mst` which produces identical topologies and is much quicker. 

See the following snippet:
```
library(patchwork)
library(dplyr)
library(igraph)
library(ape)
library(tidyr)
library(ggraph)

# generate random matrix
set.seed(12)
n <- 12 # number of points
mx <- matrix(round(runif(n^2, 1, 3)), ncol = n)
ix <- sample(1:n, size = n, replace = TRUE)
mx <- mx[ix, ix] # generate zero distances by recycling a few cases

# generate dist object
d <- dist(mx, method = "manhattan")

# Ape
# generate graph from ape::mst object
mst.a <- ape::mst(d) %>%
  igraph::graph_from_adjacency_matrix(mode = "undirected")

# C++ transcript in yea13, version 0.3
mst.y <- yea13::mstC(d) %>%
  igraph::graph_from_adjacency_matrix(mode = "undirected")

# Plot
p.ape <- ggraph(mst.a, layout = "igraph", algorithm = "kk") +
  geom_edge_fan(color = "grey70", label_colour = "grey30") +
  geom_node_point(size = 7, color = "grey70", fill = "grey90", shape=21) +
  scale_color_discrete(guide = F) +
  geom_node_text(aes(label = name)) +
  ggtitle("ape::mst") +
  theme_graph()

p.yea13 <- ggraph(mst.y, layout = "igraph", algorithm = "kk") +
  geom_edge_fan(color = "grey70", label_colour = "grey30") +
  geom_node_point(size = 7, color = "grey70", fill = "grey90", shape=21) +
  scale_color_discrete(guide = F) +
  geom_node_text(aes(label = name)) +
  ggtitle("yea13::mstC") +
  theme_graph()

print(p.ape + p.yea13)
```
![apemstcpp](https://user-images.githubusercontent.com/152780/91728481-6d0d3480-eba3-11ea-9cb8-0246faedfa8e.png)

`yea13:mstC` appears to be approx. 40 times quicker than `ape::mst`:

```
bench::mark(ape::mst(d), yea13::mstC(d)) %>% select(expression, min, median) 
# A tibble: 2 x 3   
expression          min   median   
<bch:expr>     <bch:tm> <bch:tm> 
1 ape::mst(d)      4.32ms   5.34ms 
2 yea13::mstC(d)  98.17µs 125.38µs
```

Although this doesn't explain different topologies produced by `igraph` and `ape`  the significant speed improvements by `yea13::mstC` warrant the closure of this issue.

**Comment by joheli** [comment]
Btw I have renamed `yea13::mstC` to `yea13::mst` as of version 0.5.
